File: src/main/java/com/xin/replace/bootstrap_class/BootstrapChangeClass.java
package com.xin.replace.bootstrap_class;

import com.xin.base.GeneralClassAdapter;
import com.xin.replace.base.BaseChangeClass;
import com.xin.replace.base.BasicChangeMethod;
import com.xin.util.InvokeUtil;
import org.objectweb.asm.MethodVisitor;

import java.lang.instrument.Instrumentation;

import static org.objectweb.asm.Opcodes.ALOAD;
import static org.objectweb.asm.Opcodes.ARETURN;
import static org.objectweb.asm.Opcodes.ASTORE;

/**
 * @author linxixin@cvte.com
 * @since 1.0
 */
public class BootstrapChangeClass extends BaseChangeClass {
    public BootstrapChangeClass(Instrumentation instrumentation, ClassLoader classLoader) {
        super(instrumentation, classLoader);
    }

    @Override
    protected String getClassName() {
        return "org.apache.catalina.startup.Bootstrap";
    }

    @Override
    protected GeneralClassAdapter getGeneralClassAdapter() {
        return new GeneralClassAdapter() {
            @Override
            public MethodVisitor visitMethod(int access, String name, String descriptor, String signature, String[] exceptions) {
                MethodVisitor mv = super.visitMethod(access, name, descriptor, signature, exceptions);

                // 当是sayName方法是做对应的修改
                if ("createClassLoader".equals(name)
                        && "(Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/ClassLoader;".equals(
                        descriptor)) {
                    System.out.println("!!@");

                    return new ChangeMethodAdapter(mv);
                } else {
                    return mv;
                }
            }
        };

    }

    public static class ChangeMethodAdapter extends BasicChangeMethod {

        public ChangeMethodAdapter(MethodVisitor methodVisitor) {
            super(methodVisitor);
        }

        @Override
        public void visitInsn(int opcode) {
            if (opcode == ARETURN) {
                mv.visitVarInsn(ASTORE, 6);
                mv.visitVarInsn(ALOAD, 6);
                mv.visitVarInsn(ALOAD, 2);
                InvokeUtil.invokeStaticMethod22(mv, BootstrapChangeClass.class, Handler::handle2);

                mv.visitVarInsn(ALOAD, 6);
            }
            mv.visitInsn(opcode);
        }

    }
}

File: src/main/java/com/xin/replace/BootstrapChangeClass.kt
package com.xin.replace

import com.xin.StoreInfo
import com.xin.StoreInfo.commonClassLoader
import com.xin.base.GeneralClassAdapter
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.Opcodes.*
import java.lang.instrument.Instrumentation
import kotlin.jvm.internal.FunctionReference

/**
 *
 * @author linxixin@cvte.com
 */


class BootstrapChangeClass(instrumentation: Instrumentation, classLoader: ClassLoader) : BaseChangeClass(instrumentation, classLoader) {

    companion object Handle {
        fun handleClassLoader(name: String, classLoader: ClassLoader, parent: ClassLoader) {
            if (name == "common") {
                commonClassLoader = classLoader
                WebappLoaderChangeClass(StoreInfo.instrumentation, classLoader)
            }
        }
    }


    override fun getGeneralClassAdapter(): GeneralClassAdapter {
        return object : GeneralClassAdapter() {
            override fun visitMethod(access: Int, name: String?, desc: String?,
                                     signature: String?, exceptions: Array<String>?): MethodVisitor {
                val mv = cv.visitMethod(access, name, desc, signature, exceptions)

                // 当是sayName方法是做对应的修改
                return if (name == "createClassLoader" && desc == "(Ljava/lang/String;Ljava/lang/ClassLoader;)Ljava/lang/ClassLoader;") {
                    ChangeMethodAdapter(mv)
                } else {
                    mv
                }
            }

            // 定义一个自己的方法访问类
            internal inner class ChangeMethodAdapter(mv: MethodVisitor) : MethodVisitor(ASM4, mv) {

                override fun visitInsn(opcode: Int) {


                    if (opcode == ARETURN) {
                        mv.visitVarInsn(ASTORE, 6)
                        invokeHandle(mv, BootstrapChangeClass.Handle::handleClassLoader as FunctionReference) { mv ->
                            mv.visitVarInsn(ALOAD, 1)
                            mv.visitVarInsn(ALOAD, 6)
                            mv.visitVarInsn(ALOAD, 2)
                        }
                        mv.visitVarInsn(ALOAD, 6)
                    }
                    mv.visitInsn(opcode)
                }
            }
        }
    }

    override fun getClassName(): String {
        return "org.apache.catalina.startup.Bootstrap"
    }

    override fun toString(): String {
        return "${getClassName()} 替换类"
    }
}
